### PR TITLE
[Refactor] Decompose CIDriver into 4 SRP collaborators (combined #1179+#1289)

### LINKED ISSUE #1357: [refactor] Decompose CIDriver god-class into 4 SRP collaborators (combined #1179+#1289)
## Summary

`CIDriver` (`hephaestus/automation/ci_driver.py`) is a god-class violating the
Single Responsibility Principle (SRP). At time of filing it is 3,570 lines with
~60 methods. This issue tracks decomposing it into four narrow collaborators
using the Dependency Inversion Principle (DIP), combining the two duplicate
issues #1179 and #1289 (and their two competing implementations #1292 and
#1344) into a single tracked workstream.

## Collaborators

| Module | Responsibility |
|--------|---------------|
| `hephaestus/automation/pr_discovery.py` | Viewer-login caching + PR enumeration (issue-driven, bot, failing) |
| `hephaestus/automation/ci_check_inspector.py` | Required CI check queries (`gh pr checks`) |
| `hephaestus/automation/ci_fix_orchestrator.py` | CI fix session management (codex + claude paths) |
| `hephaestus/automation/post_merge_processor.py` | Drive-green `/learn` + `/compact` after PR merges |

## Design Constraints

- **Narrow-callable injection (DIP):** each collaborator receives only the
  specific `Callable[[], T]` providers it needs — NEVER `self`/`CIDriver`.
- **Lambda wrappers on all injected callables**, so `patch.object(driver, "_method")`
  intercepts through the lambda at call time (not captured at init). Essential
  for the existing test suite.
- **All existing `patch.object(CIDriver, '_method')` test targets preserved**
  via delegation stubs on `CIDriver`.
- **Target:** reduce `ci_driver.py` to ≤ 2,450 lines (delegation stubs only).

## Out of scope

#1179's broader audit also flagged `ImplementationPhaseRunner` (2,633 ln) and 6
other >800-line files (`loop_runner.py`, `address_review.py`, `pr_reviewer.py`,
`fleet_sync.py`, `implementer.py`); those are OUT OF SCOPE for this issue —
this is CIDriver only.

## Preservation note

The decomposition must be re-implemented against current `main` (3,570 lines),
which grew past the two reference PRs via #1353–#1356. The corner-case fix-flow
methods `_resolve_dirty_pr` (#1347), `_resolve_blocked_pr` /
`_address_threads_once` (#1356), and the `resolve_dirty` recursion guard on
`_recheck_and_arm_after_fix` (#1355) MUST be preserved with exact behavior.

Refs #1179
Refs #1289

### PR DESCRIPTION
**Pull request by mvillmow**
## Summary

Decomposes the `CIDriver` god-class (3,570 lines on current `main`) into four
narrow single-responsibility collaborators, wired by the Dependency Inversion
Principle. `ci_driver.py` drops from **3,570 → 2,449 lines** (delegation stubs
plus the orchestration/arm-flow logic that legitimately stays on the driver).

| Module | Responsibility |
|--------|---------------|
| `hephaestus/automation/pr_discovery.py` | Viewer-login caching + PR enumeration (issue-driven, bot, failing) |
| `hephaestus/automation/ci_check_inspector.py` | Required CI check queries (`gh pr checks`) |
| `hephaestus/automation/ci_fix_orchestrator.py` | CI fix session management (codex + claude paths) |
| `hephaestus/automation/post_merge_processor.py` | Drive-green `/learn` + `/compact` after PR merges |

### Design

- **Narrow-callable injection (DIP):** each collaborator receives only the
  specific `Callable[[], T]` providers it needs — never `self`/`CIDriver`.
- **Lambda-wrapped providers:** every injected callable is a lambda over the
  driver's delegation stub, so `patch.object(driver, "_method")` intercepts
  through the lambda at call time. The merge-state lookup inside
  `PRDiscovery.list_open_prs_remaining` routes through an injected
  `pr_merge_state_provider` for exactly this reason.
- **Delegation stubs preserved:** every existing `patch.object(CIDriver,
  "_method")` test target remains a stub on `CIDriver`.

### Corner-case fix-flow preservation (do-not-regress)

The methods shipped after the two reference PRs are preserved with exact
behavior on `CIDriver`:

- `_resolve_dirty_pr` (#1347) — resolves armed-DIRTY merge conflicts in the
  post-fix re-arm path.
- `_resolve_blocked_pr` + `_address_threads_once` +
  `_BLOCKED_ADDRESS_MAX_ATTEMPTS=2` (#1356) — addresses unresolved review
  threads on green-but-BLOCKED PRs.
- `resolve_dirty: bool = True` recursion guard on
  `_recheck_and_arm_after_fix` (#1355).

Their regression suites (`TestRecheckAndArmAfterFix`, `TestResolveDirtyPr`,
`TestResolveBlockedPr`) pass unchanged.

### Supersedes

This re-implements the two competing reference PRs **#1292** and **#1344**
against current `main` (which grew to 3,570 lines via #1353–#1356). Both are
closed in favor of this combined PR.

## Test plan

- [x] `test_ci_driver.py` + author-scope + failing-pr-discovery + prs-mode +
      phase-agent-wiring suites pass.
- [x] New collaborator unit tests (`test_pr_discovery_helpers.py`,
      `test_ci_check_inspector_helpers.py`, `test_ci_fix_orchestrator_helpers.py`,
      `test_post_merge_helpers.py`) pass — 311 tests total green.
- [x] Omit-allowlist guard + orchestration smoke updated atomically with the
      4 new module paths.
- [x] F1/F2/F4 corner-case regression tests pass.
- [x] `ruff check` clean, `mypy` clean (no new errors), `pre-commit
      --all-files` all hooks pass.

Closes #1357

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: hephaestus/automation/ci_driver.py [hunk 1]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+        # providers / bound methods it needs (DIP) — never ``self``. The
+        # injected callables are wrapped in lambdas (or passed as bound methods
+        # that delegate further) so ``patch.object(driver, "_method")`` in tests
+        # still intercepts through the indirection at call time. The viewer-login
+        # cache (#821) now lives inside PRDiscovery.
+        self._pr_discovery = PRDiscovery(
+            options_provider=lambda: self.options,
+            status_tracker_provider=lambda: self.status_tracker,
+            repo_root_provider=lambda: self.repo_root,
+            pr_merge_state_provider=lambda pr_number: self._pr_merge_state(pr_number),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:154]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 2]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+        )
+        self._check_inspector = CICheckInspector(
+            get_pr_branch=lambda pr_number: self._get_pr_branch(pr_number),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:157]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 3]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+            options_provider=lambda: self.options,
+        )
+        self._fix_orchestrator = CIFixOrchestrator(
+            options_provider=lambda: self.options,
+            repo_root_provider=lambda: self.repo_root,
+            state_dir_provider=lambda: self.state_dir,
+            status_tracker_provider=lambda: self.status_tracker,
+            get_pr_branch=lambda pr_number: self._get_pr_branch(pr_number),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:165]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 4]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+            get_worktree_path=lambda issue_number, pr_number: self._get_worktree_path(
+                issue_number, pr_number
+            ),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:168]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 5]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+            format_review_threads_block=lambda pr_number: self._format_review_threads_block(
+                pr_number
+            ),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:171]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 6]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+            failing_required_check_names=lambda pr_number: self._failing_required_check_names(
+                pr_number
+            ),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:174]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 7]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+        )
+        self._post_merge = PostMergeProcessor(
+            options_provider=lambda: self.options,
+            repo_root_provider=lambda: self.repo_root,
+            get_worktree_path=lambda issue_number, pr_number: self._get_worktree_path(
+                issue_number, pr_number
+            ),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:181]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 8]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+            load_arming_state=lambda issue_number: self._load_arming_state(issue_number),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:182]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.

File: hephaestus/automation/ci_driver.py [hunk 9]
```diff
@@ -146,11 +139,51 @@ def __init__(self, options: CIDriverOptions) -> None:
...
+            save_arming_state=lambda issue_number, record: self._save_arming_state(
+                issue_number, record
+            ),
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply hephaestus/automation/ci_driver.py:185]: Intentional. This lambda is a Dependency-Inversion injection wrapper (issue #1357): it defers the `self._method` lookup to call time so `patch.object(driver, '_method')` intercepts correctly in the test suite. Inlining to a bare method reference would capture the bound method at init and break every test that patches these collaborators. Keeping as-is by design.
